proposals/distributed-async-epoch-reward.md

```diff
@@ -0,0 +1,294 @@
+maximum depth of the history is chosen to be the same as the depth of stake
+history for consistency, which is 512.
+
+The layout of `EpochRewardHistory` sysvar is shown in the following pseudo code.
+
+```
+struct RewardHistoryEntry {
+   total_reward_in_lamport: u64,          // total rewards for this epoch
+   distributed_reward_in_lamport: u64,    // already distributed reward amount
+   root_hash: Option<Hash>,               // hash computed from all EpochRewardReserves
+}
+
+type EpochReward = (Epoch, RewardHistoryEntry);
+
+struct RewardHistory {
+   rewards: [EpochReward; 512],
+}
+```
+
+The `EpochRewardHistory` is updated at the start of the first block of the
+epoch (before any transactions are processed), as both the total epoch rewards
+and vote account rewards become available at this time. Once the vote account
+rewards have been distributed, a `RewardHistoryEntry` will be added to the
+`EpochRewardHistory` array with the total amount of rewards and the amount
+distributed, which is equal to the sum of the vote account rewards. If the
+length of the history array exceeds the maximum depth (512), the oldest entries
+will be removed from the array (i.e., "popped").
+
+The `root_hash` is the hash of all the `EpochRewardReserve` accounts hashes.
+Initially, at the start of the epoch's first block, the `root_hash` is set to
+None. Once all the `EpochRewardReserve` accounts are filled and the reward
+computation phase is complete, the `root_hash` is calculated by hashing all the
+`EpochRewardReserve` hashes.
+
+```
+root_hash = Hash(EpochRewardReserve[..].reserve_hash)
+
+
+
+                     +-------------+
+                     |             |
+                     |  root_hash  |
+                     |             |
+                     +------+------+
+                            |
+    +--------------+--------+-----------+
+    |              |                    |
++---+----+   +-----+--+             +---+-----+
+|reserve |   | reserve|   ......    | reserve |
+| hash   |   |  hash  |             |  hash   |
++--------+   +--------+             +---------+
+
+```
+
+### `EpochRewardReserve` Sysvar Account
+
+`EpochRewardReserve` sysvar account tracks the reward to be distributed for one
+particular block in the `reward distribution phase`.
+
+The address for a particular `EpochRewardReserve` account is derived from the
+base Sysvar RewardReserve address and u64 block height for the corresponding reward
+distribution block during the reward distribution phase, i.e.
+`address = hash("SysvarRewardReserve11111111111111111111111111111", block_height)`.
+
+
+```
+crate::declare_sysvar_id!("SysvarRewardReserve11111111111111111111111111111", EpochRewardReserve);
+
+struct EpochRewardReserve {
+   reserve_hash: Hash,
+}
+```
+
+The account balance of `RewardsReserve` is equal to the total amount (in
+lamports) of all rewards to be distributed in a block. The `reserve_hash` is
```

Comment:
how will we handle the case where an `EpochRewardReserve` account already exists as a system account. this would occur should someone deposit some lamports there prior to the Epoch Rewards Period.  similarly should someone deposit lamports into the account _during_ the Epoch Rewards Period, leaving an excess balance after all anticipated rewards have been distributed.

i think the only thing that makes sense is to shadow the account's lamports field with an `epoch_rewards_lamports` field in the account data that runtime fully controls.  we c an autopromote the system account to a sysvar account at creation and burn any excess lamports at the end of distribution, so long as `epoch_rewards_lamports` is zero.